contributors/design-proposals/federated-statefulsets.md

```diff
@@ -0,0 +1,491 @@
+additional LB Services** later in this doc.
+
+The trick for individual pods being able to discover this additional dns identity through the LB
+service is the federated controller first creating SRV records for each stateful pod in the
+federation dns server against the governing service domain
+```store.mynamespace.myfederation.svc.example.com``` and then later updating an A record of the IP
+resolved for the specific LB service (for example ```store-c1-0.mynamespace.myfederation.svc.example.com```)
+against each of these SRV records.
+A peer pod in a stateful set then can discover its federated peers by doing a dns query against
+the governing service domain ```store.mynamespace.myfederation.svc.example.com```.
+It should be noted that as of now the federation service controller ignores service records which
+are not of type Loadbalancer, and does not update them in the federation dns server. This behaviour
+will be aptly modified to ensure that the dns records against the headless services are also updated
+in the federated dns server if the need be. The federation service controller will determine this based
+on the user annotation (```federation.kubernetes.io/statefulset-lb-needed```) with value ```true``` present
+on the federated headless service. Please not that this is the same annotation which is used on the
+federated statefulset to determine, if this statefulset needs LB services or not. The federation
+statefulset controller will update this on the federated headless service while creating it, or if
+its applied/updated later, will update it accordingly.
+Additionally the information about the stateful replicas (and the same name LB services) for which the
+service controller should update the dns records against the headless service will also need to be passed
+on to the headless service. The proposal is to pass this information through annotations on the headless
+service as below:
+```
+federation.kubernetes.io/stateful-replica-list: <replica-name-1>, <replica-name-2>, ..., <replica-name-n>
+```
+The federation service controller, if finds the names in this list against a federated headless service,
+will look for matching federated LB services and update the CNAME record for the headless service with the
+relevant SRV records for all the LB services in the federation dns server.
+This list can certainly be updated to the first class API data, if the federation API objects and clients
+are updated and maintained separately to have additional fields/data then that present in k8s API.
+
+Once this happens, the stateful pods can discover peers by querying the governing service domain (federated
+headless service). Taking the same example listed earlier if the peer (or a user) queries the service
+```store.mynamespace.myfederation.svc.example.com```, the below SRV records will be returned.
+```
+store-c1-0.etcd.mynamespace.myfederation.svc.example.com
```

Comment:
Have added some text relating the storage at the location describing the naming in "Instance identity and discovery (local identities)".